libraries/ms-helper/src/persist.c

```diff
@@ -0,0 +1,158 @@
+#include "persist.h"
+#include <inttypes.h>
+#include "log.h"
+#include "soft_timer.h"
+// The persistance layer allocates one page of flash so we can erase the entire page when full.
+// To reduce the number of erases, we partition the page into a number of sections. Each section
+// is the size of the specified blob plus a header. The header is used to mark the section as valid
+// and record the blob's size.
+//
+// By default, all sections are invalid. Only one section should be valid at a time. At each write,
+// all previous sections are marked as invalid and the new section is considered valid.
+//
+// Section format: [ marker (u32) | size (u32) | blob (u32 * n) ]
+// We initially write to the size field and blob data, skipping the marker. We then consider the
+// default unwritten value of marker as "valid" (0xFFFFFFFF). To invalidate the section, we write
+// 0x0 to the marker field. Because we skipped it earlier, this write succeeds and does not require
+// erasing the page. If all sections are invalid, we erase the entire page.
+//
+// At init, the persistance layer attemps to load the blob with the data stored in flash.
+
+// Erased flash defaults to all 1's
+#define PERSIST_VALID_MARKER 0xFFFFFFFF
+#define PERSIST_INVALID_SIZE 0xFFFFFFFF
+#define PERSIST_INVALID_ADDR UINTPTR_MAX
+#define PERSIST_BASE_ADDR FLASH_PAGE_TO_ADDR(PERSIST_FLASH_PAGE)
+#define PERSIST_END_ADDR (PERSIST_BASE_ADDR + FLASH_PAGE_BYTES)
+
+typedef struct PersistHeader {
+  uint32_t marker;
+  uint32_t size_bytes;
+} PersistHeader;
+
+static void prv_periodic_commit(SoftTimerID timer_id, void *context) {
+  PersistStorage *persist = context;
+
+  if (persist->prev_flash_addr == PERSIST_INVALID_ADDR) {
+    // This is the first commit for a new page
+    persist_commit(persist);
+  } else {
+    // We should check if our data has changed from the stored copy
+    uint32_t buffer = 0;
+    PersistHeader header = { 0 };
+    flash_read(persist->prev_flash_addr, sizeof(header), (uint8_t *)&header, sizeof(header));
+    if (persist->blob_size != header.size_bytes) {
+      // Wrong size - make sure we reflect the new size
+      persist_commit(persist);
+    } else {
+      uintptr_t addr = persist->prev_flash_addr + sizeof(header);
+      uint32_t *blob_u32 = (uint32_t *)persist->blob;
+      for (size_t i = 0; i < header.size_bytes / sizeof(buffer); i++, addr += sizeof(buffer)) {
+        flash_read(addr, sizeof(buffer), (uint8_t *)&buffer, sizeof(buffer));
+        if (blob_u32[i] != buffer) {
+          // There is a difference between the two copies, so commit it
+          persist_commit(persist);
+          break;
+        }
+      }
+    }
+  }
+
+  soft_timer_start_millis(PERSIST_COMMIT_TIMEOUT_MS, prv_periodic_commit, persist, NULL);
+}
+
+StatusCode persist_init(PersistStorage *persist, void *blob, size_t blob_size) {
+  if (blob_size > FLASH_PAGE_BYTES) {
+    return status_code(STATUS_CODE_OUT_OF_RANGE);
+  } else if (blob_size % FLASH_WRITE_BYTES != 0) {
```

Comment:
I wanted to keep it aligned because we can't actually pad the value - we'd really be copying in additional data after the blob, which felt really wrong to me.